toc.md

```diff
@@ -428,77 +429,234 @@ An additional backlog of 936 hours of image processing jobs everyday, along with
+
 ![parallelism_workeholic_diagram](/images/parallelism_workeholic_diagram.png)
 
-The diagram above shows how Workerholic uses multiple processes if we had the CPU cores available to do so. The OS scheduler schedules the cores to each process, each process has its own worker threads which can poll jobs from Redis and work on them. So if we have four CPU cores we can have computational resources allocated to four different processes potentially at the same time, effectively allowing them to run in parallel, reducing the enqueuing:processing ratio from 1:40 to 1:10.
+Here is the code we wrote to make the above model possible:
+
+```ruby
+module Workerholic
+  class Starter
+    # ...
+
+    def self.launch
+      fork_processes if options[:processes] && options[:processes] > 1
+
+      Workerholic.manager = Manager.new(auto_balance: options[:auto_balance])
+      Workerholic.manager.start
+    end
+
+    def self.fork_processes
+      (options[:processes] - 1).times do
+        PIDS << fork do
+          Workerholic.manager = Manager.new(auto_balance: options[:auto_balance])
+          Workerholic.manager.start
+        end
+      end
+
+      PIDS.freeze
+    end
+
+    # ...
+  end
+end
+```
+
+We fork processes if the user specified the option when starting Workerholic. Then we fork the number of processes specified minus one since the main process will serve as one of these processes, effectively having 1 parent process and N - 1 children processes. We make sure to store the processes ids `pid`, inside a `PIDS` constant for future reference. In each of these children processes we start the manager, which starts all the workers and the other components such as the Job Scheduler. Once these children processes are started we start the manager in the parent process.
+
+In the next section we will look at how forking children processes impacts the consumption of the available memory.
 
 ##### Processes and Memory consumption
-We're able to create more processes, but surely that comes with a cost like everything else right?
+
+A Ruby process can fork multiple children processes. The memory model for a process is as follows:
 
 ![efficiency_processes_memory_design](/images/efficiency_processes_memory_design.png)
 
-Before we get into that, let's talk a little bit more about processes. Each process has its own address space, stack, and heap. When you fork a process in Ruby, you create a child process which will get its own stack and share the same heap initially. This is called copy-on-write, meaning that the child process will share the same resources with the parent process, until modifications are made to that resource, in which case, that resource will be written into the child's own heap.
+Each process has its own address space, stack, and heap. When a process is forked in Ruby, a child process is created, which will get its own stack. As for the heap, depending on the OS the copy-on-write mechanism is employed:
```

Comment:
Each process has its own address space, stack, and heap. When a process is forked, a child process is created with its own stack. As for the heap, depending on the OS the copy-on-write mechanism might be employed: